### PR TITLE
[CodeGen] change prototype of RegClassFilterFunc

### PR DESCRIPTION
**Pull request by cmc-rep**
change prototype of RegClassFilterFunc so that we can filter not just by RegClass. We need to implement more complicated filter based upon some other info associated with each register.

### CODE REVIEW

File: llvm/include/llvm/CodeGen/RegAllocCommon.h [hunk 1]
```diff
@@ -13,19 +13,21 @@
...
-                           const TargetRegisterClass &RC)> RegClassFilterFunc;
+                           const MachineRegisterInfo &MRI, const Register &Reg)>
+    RegClassFilterFunc;
 
 /// Default register class filter function for register allocation. All virtual
 /// registers should be allocated.
 static inline bool allocateAllRegClasses(const TargetRegisterInfo &,
-                                         const TargetRegisterClass &) {
+                                         const MachineRegisterInfo &,
+                                         const Register &) {
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/RegAllocCommon.h:28]: Register should be passed by value

**Review by arsenm — COMMENTED** [review]
Should link to additional context in usage patch

File: llvm/include/llvm/CodeGen/RegAllocCommon.h (hunk 1 above)
Comment (cmc-rep) [reply llvm/include/llvm/CodeGen/RegAllocCommon.h:28]: will do

Comment (cmc-rep) [reply llvm/include/llvm/CodeGen/RegAllocCommon.h:28]: I tried it, pass-by-value does not work in this case. With MSVC, I am getting the following error:

error C2027: use of undefined type 'llvm::Register'

It looks like this header file does not want any dependence on other header file

Comment (cmc-rep) [reply llvm/include/llvm/CodeGen/RegAllocCommon.h]: adding llvm/CodeGen/Register.h seems okay. fixed the error

**Comment by cmc-rep** [comment]
messed up the merge

**Pull request closed without merging**